btree/btree.c

```diff
@@ -3875,16 +3875,11 @@ static int64_t btree_put_kv_tick(struct m0_sm_op *smop)
 			}
 		}
 		/** Fall through if path_check is successful. */
-	case P_DECIDE_NXT: {
-		if (bop->bo_opc == M0_BO_PUT) {
-			if (oi->i_key_found)
-				return P_ACT;
-			else
-				return  P_MAKESPACE;
-		} else {
-			M0_ASSERT(bop->bo_opc == M0_BO_UPDATE);
+	case P_SANITY_CHECK: {
+		if (bop->bo_opc == M0_BO_PUT && !oi->i_key_found)
+			return  P_MAKESPACE;
+		else
```

Comment:
It is necessary to have checks in P_ACT, as in  P_SANITY_CHECK we can not call fail routine directly as callback needs to be done first. and callback will be called in P_ACT.

Reply:
If the check fails and we can return the proper error code then I am thinking we do not need the callback. I understand we can provide an error status in the callback but maybe we can just skip this part - I think it will be more useful for the caller to debug his code as he can check for the return value of the function instead of looking for the error in the callback. I hope these are not a lot of changes.

Reply:
Yes we can do that , but currently ut is is checking callback flag which will get set in callback routine only, If we do above changes, we may need to change error handling in ut. 
Also, it can happen that user request to put the key which is already exists, in such case don't we need to do callback , which will tell user that key already exists.